Objects/genericaliasobject.c

```diff
@@ -244,6 +247,48 @@ make_parameters(PyObject *args)
     return parameters;
 }
 
+/* If obj is a generic alias, substitute type variables params
+   with substitutions argitems. */
+static PyObject *
+subs_tvars(PyObject *obj, PyObject *params, PyObject **argitems)
+{
+    _Py_IDENTIFIER(__parameters__);
+    PyObject *subparams;
+    if (_PyObject_LookupAttrId(obj, &PyId___parameters__, &subparams) < 0) {
+        return NULL;
+    }
+    if (subparams && PyTuple_Check(subparams)) {
+        Py_ssize_t nparams = PyTuple_GET_SIZE(params);
+        Py_ssize_t nsubargs = PyTuple_GET_SIZE(subparams);
+        if (nsubargs != 0) {
+            PyObject *subargs = PyTuple_New(nsubargs);
+            if (subargs == NULL) {
+                Py_DECREF(subparams);
+                return NULL;
+            }
+            for (Py_ssize_t i = 0; i < nsubargs; ++i) {
+                PyObject *arg = PyTuple_GET_ITEM(subparams, i);
+                Py_ssize_t iparam = tuple_index(params, nparams, arg);
+                if (iparam >= 0) {
+                    arg = argitems[iparam];
+                }
+                Py_INCREF(arg);
+                PyTuple_SET_ITEM(subargs, i, arg);
+            }
+            obj = PyObject_GetItem(obj, subargs);
```

Comment:
Since this is the operation we've been preparing for, I'd set it apart with a blank line before and after.